### PR TITLE
Sync back after upstreaming

### PR DESCRIPTION
**Pull request by banach-space**
The following PR's are merged back:
* https://reviews.llvm.org/D113747
* https://reviews.llvm.org/D113666
* https://reviews.llvm.org/D113395
* https://reviews.llvm.org/D113571
* https://reviews.llvm.org/D113569
* https://reviews.llvm.org/D113090
* https://reviews.llvm.org/D113278
* https://reviews.llvm.org/D113023

### CODE REVIEW

**Review by jeanPerier — APPROVED** [review]
LGTM and builds OK for me, thanks !
I'll wait tomorrow to merge it to wait for potential other reviews.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 1]
```diff
@@ -20,13 +20,10 @@
 #include "flang/Optimizer/Dialect/FIRAttr.h"
 #include "flang/Optimizer/Dialect/FIRDialect.h"
 #include "flang/Optimizer/Dialect/FIROps.h"
-#include "flang/Optimizer/Dialect/FIRType.h"
-#include "flang/Optimizer/Support/FIRContext.h"
 #include "flang/Optimizer/Support/InternalNames.h"
 #include "flang/Optimizer/Support/KindMapping.h"
 #include "flang/Optimizer/Support/TypeCode.h"
 #include "mlir/Conversion/ArithmeticToLLVM/ArithmeticToLLVM.h"
-#include "mlir/Conversion/LLVMCommon/TypeConverter.h"
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/CodeGen/CodeGen.cpp:29]: It's not clear why these are being removed. Are they no longer needed? If not, why not?

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 2]
```diff
@@ -2606,23 +2614,24 @@ struct GlobalOpConversion : public FIROpConversion<fir::GlobalOp> {
...
+    // fir.ref<fir.box> is actually a no op in LLVM.
     if (load.getType().isa<fir::BoxType>()) {
-      rewriter.replaceOp(load, operands[0]);
+      rewriter.replaceOp(load, adaptor.getOperands()[0]);
     } else {
-      auto ty = convertType(load.getType());
-      auto at = load->getAttrs();
-      rewriter.replaceOpWithNewOp<mlir::LLVM::LoadOp>(load, ty, operands, at);
+      mlir::Type ty = convertType(load.getType());
+      ArrayRef<NamedAttribute> at = load->getAttrs();
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/CodeGen/CodeGen.cpp:2632]: `llvm::ArrayRef`

There are other classes named ArrayRef in flang, so it's good to use the namespace.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp [hunk 3]
```diff
@@ -2954,60 +2971,71 @@ struct ZeroOpConversion : public FIROpConversion<fir::ZeroOp> {
...
       auto c0 = mlir::ArrayAttr::get(ctx, rewriter.getI32IntegerAttr(0));
       ptr = rewriter.create<mlir::LLVM::ExtractValueOp>(loc, ty, ptr, c0);
     }
-    auto c0 = genConstantIndex(isPresent.getLoc(), idxTy, rewriter, 0);
+    mlir::LLVM::ConstantOp c0 =
+        genConstantIndex(isPresent.getLoc(), idxTy, rewriter, 0);
     auto addr = rewriter.create<mlir::LLVM::PtrToIntOp>(loc, idxTy, ptr);
     rewriter.replaceOpWithNewOp<mlir::LLVM::ICmpOp>(
         isPresent, mlir::LLVM::ICmpPredicate::ne, addr, c0);
+
```
Comment (schweitzpgi) [inline flang/lib/Optimizer/CodeGen/CodeGen.cpp:3016]: flang code tends toward being compact and not using blank lines.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp (hunk 1 above)
Comment (clementval) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:29]: I think they have been moved to the `TypeConverter.h`

Comment (banach-space) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:29]: This change is from: https://reviews.llvm.org/rGf1a2b507896c3fec8b725e489d9c0a9816d292c0

Comment (banach-space) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:29]: Let me expand - these headers were missing from `TypeConverter.h` (so that `TypeConverter.h` includes everything that it needs). From what we could tell, these are not needed in `CodeGen.cpp`.

File: flang/lib/Optimizer/CodeGen/CodeGen.cpp (hunk 3 above)
Comment (banach-space) [reply flang/lib/Optimizer/CodeGen/CodeGen.cpp:3016]: Sorry about that. I do tend to inject empty lines as that makes code easier for me to read. I'm not sure whether that's my poor vision (I don't get 20/20 even with prescription glasses) or how my brain is wired up.

We can revert this, that's not a problem. But I would really prefer to make such changes on `main` and then sync them back in into `fir-dev` (or perhaps skip that final step altogether). In fact, IMO it would be fine to just submit such an update to `main` as an NFC (without a review). Perhaps once all of this file has been upstreamed?

Basically, I'd like to minimise the traffic between `fir-dev` and `main`.

**Comment by jeanPerier** [comment]
Hi @banach-space, sorry for not coming back sooner, LGTM, could you solve the conflict so that I can merge it ?

**Comment by banach-space** [comment]
> Hi @banach-space, sorry for not coming back sooner, LGTM, could you solve the conflict so that I can merge it ?

@jeanPerier No worries! I've just rebased.
